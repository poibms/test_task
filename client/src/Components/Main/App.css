@import "../Header/Header.css";

* {
    padding: 0;
    margin: 0;
  }
  
  body{
    margin: 0;
    padding: 20px 0;
    box-sizing: border-box;
    background-color: #f2f2f2;
    font-family: 'Roboto', sans-serif;
  }
  
  .wrapper {
    max-width: 1200px;
    margin: 0 auto;
    background-color: #fff;
    padding: 0 15px;
    border-radius: 10px;
    min-height: 525px;
  }
  
  .content {
    padding: 20px 0;
  }
  
.search-panel {
  padding: 30px 0;
  display: flex;
  justify-content: center;
}

  