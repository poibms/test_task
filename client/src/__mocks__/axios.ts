// const axios = {
// 	get: jest.fn(() => Promise.resolve({ data: {} })),
// 	create: () => axios,
// 	defaults: {
// 		adapter: {},
// 	},
// };

const axios: any = jest.genMockFromModule('axios');

axios.create.mockReturnThis();

export default axios;
