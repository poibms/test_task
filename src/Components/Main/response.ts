export const error = {
	cod: '404',
	message: 'city not found',
};

export const data = {
	cod: '200',
	message: 0,
	cnt: 5,
	list: [
		{
			dt: 1638284400,
			main: {
				temp: -1.45,
				feels_like: -8.45,
				temp_min: -1.45,
				temp_max: -1.41,
				pressure: 973,
				sea_level: 973,
				grnd_level: 955,
				humidity: 91,
				temp_kf: -0.04,
			},
			weather: [
				{ id: 600, main: 'Snow', description: 'light snow', icon: '13n' },
			],
			clouds: { all: 100 },
			wind: { speed: 9.22, deg: 233, gust: 16.93 },
			visibility: 657,
			pop: 1,
			snow: { '3h': 0.91 },
			sys: { pod: 'n' },
			dt_txt: '2021-11-30 15:00:00',
		},
		{
			dt: 1638295200,
			main: {
				temp: -1.75,
				feels_like: -8.75,
				temp_min: -2.36,
				temp_max: -1.75,
				pressure: 975,
				sea_level: 975,
				grnd_level: 960,
				humidity: 89,
				temp_kf: 0.61,
			},
			weather: [
				{ id: 600, main: 'Snow', description: 'light snow', icon: '13n' },
			],
			clouds: { all: 100 },
			wind: { speed: 9.37, deg: 246, gust: 16.6 },
			visibility: 554,
			pop: 1,
			snow: { '3h': 0.82 },
			sys: { pod: 'n' },
			dt_txt: '2021-11-30 18:00:00',
		},
		{
			dt: 1638306000,
			main: {
				temp: -1.95,
				feels_like: -8.95,
				temp_min: -2.2,
				temp_max: -1.95,
				pressure: 978,
				sea_level: 978,
				grnd_level: 963,
				humidity: 83,
				temp_kf: 0.25,
			},
			weather: [
				{ id: 600, main: 'Snow', description: 'light snow', icon: '13n' },
			],
			clouds: { all: 100 },
			wind: { speed: 8.43, deg: 259, gust: 15.75 },
			visibility: 10000,
			pop: 0.88,
			snow: { '3h': 0.54 },
			sys: { pod: 'n' },
			dt_txt: '2021-11-30 21:00:00',
		},
		{
			dt: 1638316800,
			main: {
				temp: -2.4,
				feels_like: -8.97,
				temp_min: -2.4,
				temp_max: -2.4,
				pressure: 985,
				sea_level: 985,
				grnd_level: 967,
				humidity: 92,
				temp_kf: 0,
			},
			weather: [
				{ id: 600, main: 'Snow', description: 'light snow', icon: '13n' },
			],
			clouds: { all: 100 },
			wind: { speed: 6.93, deg: 234, gust: 12.71 },
			visibility: 239,
			pop: 1,
			snow: { '3h': 0.84 },
			sys: { pod: 'n' },
			dt_txt: '2021-12-01 00:00:00',
		},
		{
			dt: 1638327600,
			main: {
				temp: -2.61,
				feels_like: -9.2,
				temp_min: -2.61,
				temp_max: -2.61,
				pressure: 986,
				sea_level: 986,
				grnd_level: 968,
				humidity: 85,
				temp_kf: 0,
			},
			weather: [
				{ id: 600, main: 'Snow', description: 'light snow', icon: '13n' },
			],
			clouds: { all: 100 },
			wind: { speed: 6.85, deg: 252, gust: 12.5 },
			visibility: 4830,
			pop: 0.8,
			snow: { '3h': 0.98 },
			sys: { pod: 'n' },
			dt_txt: '2021-12-01 03:00:00',
		},
	],
	city: {
		id: 524901,
		name: 'Moscow',
		coord: { lat: 55.7522, lon: 37.6156 },
		country: 'RU',
		population: 0,
		timezone: 10800,
		sunrise: 1638250403,
		sunset: 1638277405,
	},
};
