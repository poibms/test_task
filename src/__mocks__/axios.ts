// const axios = {
// 	get: jest.fn(() => Promise.resolve({ data: {} })),
// 	create: () => axios,
// 	defaults: {
// 		adapter: {},
// 	},
// };

const axios: any = jest.createMockFromModule('axios');

axios.create.mockReturnThis();

export default axios;
