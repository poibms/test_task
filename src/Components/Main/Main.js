import React, { Component } from 'react';
import './App.css';
import { connect } from 'react-redux';
import Header from '../Header/Header';
import SearchPanel from '../SearchPanel/SearchPanel';
import Weather from '../Weather/Weather';
import LocalStorageServices from '../../Services/LocalStorageServices';
import {
	addCurntWeather,
	addSearchHistory,
	addNearestWeather,
	removeSearchHistory,
} from '../../Actions/ActionCreator';
import { getWeatherById, getWeatherByName } from '../../Reducers/Weather';

class Main extends Component {
	componentDidUpdate(prevProps) {
		const { searchHistory } = this.props;

		if (prevProps.searchHistory !== searchHistory) {
			LocalStorageServices.addSearchHistory(JSON.stringify(searchHistory));
		}
	}

	submitRequestByName = (value) => {
		console.log(this.props);
		this.props.getByName(value, this.props);
		// await searchByName(value)
		// 	.then((response) => {
		// 		const dt = Date(response.data.dt);
		// 		const crntDate = dt.slice(0, 15);
		// 		const time = dt.slice(16, 24);
		// 		const tempArr = response.data.list;
		//
		// 		// eslint-disable-next-line no-shadow
		// 		const { nearestWeather, addCurntWeather, addNearestWeather } =
		// 			this.props;
		// 		// eslint-disable-next-line no-plusplus
		// 		for (let index = 0; index < tempArr.length; index++) {
		// 			if (index === 0) {
		// 				const newCurWeather = {
		// 					name: response.data.city.name,
		// 					country: response.data.city.country,
		// 					temp: Math.ceil(tempArr[index].main.temp),
		// 					feel: Math.floor(tempArr[index].main.feels_like),
		// 					crntDate,
		// 					time,
		// 					weather: tempArr[index].weather,
		// 				};
		// 				addCurntWeather(newCurWeather);
		// 			} else {
		// 				const newNearestWeather = {
		// 					temp: Math.ceil(tempArr[index].main.temp),
		// 					feels_like: Math.floor(tempArr[index].main.feels_like),
		// 					wind: tempArr[index].wind.speed,
		// 					time: tempArr[index].dt_txt.slice(11, 20),
		// 					id: tempArr[index].dt,
		// 				};
		//
		// 				if (nearestWeather.length === 4) {
		// 					// eslint-disable-next-line no-param-reassign
		// 					nearestWeather.splice(0, nearestWeather.length);
		//
		// 					return newNearestWeather;
		// 				}
		// 				addNearestWeather(newNearestWeather);
		// 			}
		// 		}
		//
		// 		const { searchHistory } = this.props;
		// 		const newPost = {
		// 			name: response.data.city.name,
		// 			country: response.data.city.country,
		// 			id: response.data.city.id,
		// 		};
		// 		const { id } = response.data.city;
		// 		// eslint-disable-next-line no-shadow
		// 		const { addSearchHistory } = this.props;
		// 		const elem = searchHistory.find((item) => item.id === id);
		// 		if (elem) {
		// 			const index = searchHistory.indexOf(elem);
		// 			searchHistory.splice(index, 1);
		//
		// 			const newData = [...searchHistory, newPost];
		// 			addSearchHistory(newPost);
		//
		// 			return newData;
		// 		}
		// 		const newData = [...searchHistory, newPost];
		// 		addSearchHistory(newPost);
		//
		// 		LocalStorageServices.addSearchHistory(JSON.stringify(newData));
		// 	})
		// 	.catch((e) => {
		// 		alert(wrongValue, e);
		// 	});
	};

	submitRequestById = async (id) => {
		this.props.getById(id, this.props);
		// const response = await searchById(id);
		// const dt = Date(response.data.dt);
		// const crntDate = dt.slice(0, 15);
		// const time = dt.slice(16, 24);
		//
		// const tempArr = response.data.list;
		// // eslint-disable-next-line no-shadow
		// const { nearestWeather, addCurntWeather, addNearestWeather } = this.props;
		// for (let index = 0; index < tempArr.length; index = index + 1) {
		// 	if (index === 0) {
		// 		const newCurWeather = {
		// 			name: response.data.city.name,
		// 			country: response.data.city.country,
		// 			temp: Math.ceil(tempArr[index].main.temp),
		// 			feel: Math.floor(tempArr[index].main.feels_like),
		// 			crntDate,
		// 			time,
		// 			weather: tempArr[index].weather,
		// 		};
		// 		addCurntWeather(newCurWeather);
		// 	} else {
		// 		const newNearestWeather = {
		// 			temp: Math.ceil(tempArr[index].main.temp),
		// 			feels_like: Math.floor(tempArr[index].main.feels_like),
		// 			wind: tempArr[index].wind.speed,
		// 			time: tempArr[index].dt_txt.slice(11, 20),
		// 			id: tempArr[index].dt,
		// 		};
		//
		// 		if (nearestWeather.length === 4) {
		// 			// eslint-disable-next-line no-param-reassign
		// 			nearestWeather.splice(0, nearestWeather.length);
		//
		// 			return newNearestWeather;
		// 		}
		// 		addNearestWeather(newNearestWeather);
		// 	}
		// }
		//
		// const { searchHistory } = this.props;
		// // eslint-disable-next-line no-shadow
		// const { addSearchHistory } = this.props;
		// const newPost = {
		// 	name: response.data.city.name,
		// 	country: response.data.city.country,
		// 	id: response.data.city.id,
		// };
		//
		// const elem = searchHistory.find((item) => item.id === id);
		// const index = searchHistory.indexOf(elem);
		// searchHistory.splice(index, 1);
		// const newData = [...searchHistory, newPost];
		// addSearchHistory(newPost);
		//
		// LocalStorageServices.addSearchHistory(JSON.stringify(newData));
	};

	render() {
		// eslint-disable-next-line no-shadow
		const { crntWeather, searchHistory, nearestWeather, removeSearchHistory } =
			this.props;

		return (
			<>
				<div className="wrapper">
					<div className="content">
						<Header {...this.props} />
						<div className="search-panel">
							<SearchPanel onSubmitByName={this.submitRequestByName} />
						</div>
						<div className="main">
							<Weather
								data={crntWeather}
								posts={searchHistory}
								nextWeather={nearestWeather}
								onSubmitById={this.submitRequestById}
								onDeleteItem={removeSearchHistory}
							/>
						</div>
					</div>
				</div>
			</>
		);
	}
}
export default connect(
	(state) => ({
		crntWeather: state.crntWeather,
		searchHistory: state.searchHistory,
		nearestWeather: state.nearestWeather,
	}),
	{
		addCurntWeather,
		addSearchHistory,
		addNearestWeather,
		removeSearchHistory,
		getByName: getWeatherByName,
		getById: getWeatherById,
	},
)(Main);
