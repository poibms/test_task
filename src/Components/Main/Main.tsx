import React, { useEffect } from 'react';
import './App.css';
import { useDispatch, useSelector } from 'react-redux';
import Header from '../Header/Header';
import SearchPanel from '../SearchPanel/SearchPanel';
import Weather from '../Weather/Weather';
import LocalStorageServices from '../../Services/LocalStorageServices';
import { removeSearchHistory } from '../../Actions/RemoveSearchHistory';
import { getWeatherById, getWeatherByName } from '../../Reducers/Weather';

const Main: React.FunctionComponent = () => {
	const crntWeather = useSelector((state: any) => state.crntWeather);
	const searchHistory = useSelector((state: any) => state.searchHistory);
	const nearestWeather = useSelector((state: any) => state.nearestWeather);
	const dispatch = useDispatch();

	const stateObj = { nearestWeather, searchHistory };

	useEffect(() => {
		if (searchHistory !== undefined) {
			LocalStorageServices.addSearchHistory(JSON.stringify(searchHistory));
		}
	});

	const submitRequestByName = async (value: any) => {
		dispatch(getWeatherByName({ value, stateObj }));
	};

	const submitRequestById = async (id: any) => {
		dispatch(getWeatherById({ id, stateObj }));
	};

	const removeSearchItem = (id: any) => {
		const array = searchHistory.filter((item: any) => item.id !== id);
		dispatch(removeSearchHistory(array));
	};

	return (
		<>
			<div className="wrapper">
				<div className="content">
					<Header />
					<div className="search-panel">
						<SearchPanel onSubmitByName={submitRequestByName} />
					</div>
					<div className="main">
						<Weather
							data={crntWeather}
							posts={searchHistory}
							nextWeather={nearestWeather}
							onSubmitById={submitRequestById}
							onDeleteItem={removeSearchItem}
						/>
					</div>
				</div>
			</div>
		</>
	);
};
export default Main;
