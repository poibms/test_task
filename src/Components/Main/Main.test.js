import React from 'react';
import { createStore } from 'redux';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { combineReducers } from '@reduxjs/toolkit';
import userEvent from '@testing-library/user-event';
import axios from 'axios';
import {
	crntWeather,
	searchHistory,
	nearestWeather,
} from '../../Reducers/Weather';
import Main from './Main';

jest.mock('axios');

const rootReducer = combineReducers({
	searchHistory,
	crntWeather,
	nearestWeather,
});

const renderWithRedux = (
	component,
	{ initialState, store = createStore(rootReducer, initialState) } = {},
) => {
	return {
		...render(<Provider store={store}>{component}</Provider>),
		store,
	};
};

const testObj = {
	name: 'Minsk',
	country: 'BY',
	temp: 13,
	feel: 10,
	crntDate: '11-22-2021',
	time: '12:41',
	weather: [
		{
			description: 'overcast clouds',
			icon: '04d',
			id: 804,
			main: 'Clouds',
		},
	],
};
const testWeather = [
	{
		feels_like: -2,
		id: 1637593200,
		temp: 1,
		time: '15:00:00',
		wind: 2.11,
	},
	{
		feels_like: -3,
		id: 1637604000,
		temp: 1,
		time: '18:00:00',
		wind: 2.11,
	},
];
const testSearchHistory = [
	{
		country: 'BY',
		id: 629634,
		name: 'Brest',
	},
	{
		country: 'BY',
		id: 625144,
		name: 'Minsk',
	},
];

describe('testing Main component', () => {
	it('render Main snapshot', () => {
		const { asFragment } = renderWithRedux(<Main />);
		expect(asFragment(<Main />)).toMatchSnapshot();
	});

	// it('testing search function', () => {
	// 	axios.get.mockImplementationOnce(() => Promise.resolve({ data: testObj }));
	// 	const { getByText, getByPlaceholderText } = renderWithRedux(<Main />);
	// 	userEvent.type(getByPlaceholderText('Enter city'), 'Minsk');
	// 	userEvent.click(getByText('Search'));
	// 	// const items = await getByText(/11-22-2021/i);
	// 	// expect(items).toBeInTheDocument();
	// 	expect(axios.get).toHaveBeenCalledTimes(1);
	// });
});

describe('testing Main component with Redux', () => {
	it('render component with initialState', () => {
		const { getByText } = renderWithRedux(<Main />, {
			initialState: {
				crntWeather: testObj,
			},
		});
		expect(getByText('Clouds')).toBeInTheDocument();
	});
	it('testing search by id through Redux', () => {
		const { getByText } = renderWithRedux(<Main />, {
			initialState: {
				searchHistory: testSearchHistory,
			},
		});
		userEvent.click(getByText(/Brest/i));
		screen.debug();
	});
});
