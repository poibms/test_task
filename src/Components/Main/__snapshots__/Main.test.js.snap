// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Main component render Main 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <div
      class="content"
    >
      <div
        class="header"
      >
        <h1>
          Weather app
        </h1>
        <div>
          <button
            class="login-button"
            type="button"
          >
            Log Out
          </button>
        </div>
      </div>
      <div
        class="search-panel"
      >
        <form
          class="search-form"
        >
          <input
            class="onChange"
            placeholder="Enter city"
            type="text"
            value=""
          />
          <input
            class="onSearch"
            type="submit"
            value="Search"
          />
        </form>
      </div>
      <div
        class="main"
      >
        <div
          class="main-content"
        >
          <div
            class="main-list"
          >
            <h3>
              Search history
            </h3>
            <ul
              class="list-item"
            />
          </div>
          <div
            class="main-weather"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
